src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -1991,14 +2042,29 @@ internal bool TrySkipBytes(int num)
 
         internal void SetSnapshot()
         {
-            _snapshot = new StateSnapshot(this);
-            _snapshot.Snap();
+            StateSnapshot snapshot = _snapshot;
+            if (snapshot is null)
+            {
+                snapshot = Interlocked.Exchange(ref _cachedSnapshot, null) ?? new StateSnapshot();
+            }
+            else
+            {
+                snapshot.Clear();
+            }
+            _snapshot = snapshot;
+            _snapshot.Snap(this);
```

Comment:
It's there so there is no intermediate state where `_snapshot` is not null but is not a correctly setup snapshot. It sets it up locally and then assigns it back to the shared variable. It could be an internlocked assignment but that isn't seems necessary.